### DISCUSSION TITLE
How strict should types be?

### DISCUSSION THREAD

**Discussion by ctron** [patternfly-yew/patternfly-yew#46] [Ideas]
I wanted to  discuss a topic on how much the Rust type system should be used to enforce things. Right now, some components to require certain child types. Some because they need access to their properties (which can be questioned) some just to ensure the "right" components get added as children.

I am talking about this latter case.

We already had the issue that a child type was not enforced, and it caused some rendering issues. So, the Rust type system could have helped.

However, there are other cases where components don't exists, or people might just want to throw in random HTML (maybe with patternfly styles/classes) which isn't possible in such cases.

So the question is: in cases where it isn't strictly necessary for the parent to "know" the child, should it enforce it's type?

@jbtrystram @phill-85 @collinsmarra I would appreciate your thoughts in this too.

**Comment by jbtrystram** [comment]
One of the strong arguments of Rust is being as strict as possible with types, and that works well to reduce bugs. 
You can even use `type` to set your own aliases and go one step further. 

I think we should stick to that philosophy wherever is possible. 
However the random HTML use case is legit as well 
Maybe we could make it more explicit ? 
``` rust
enum Child {
  Componemt(TheTypeYouWantEnforce),
  UseAtYourOwnRisk(Html)
}
```

That way you can't use some other not intended Component, except if you know what you are doing ?

**Reply by ctron** [reply]
I like the idea and I tried it out on `main` for the new `Card` implementation (in `next`). There is a `Raw` component which can be used. But only if a component actively supports this.

I considered using `Unsafe`, but that might trigger people too much. Although I think it kind of makes sense. So that name (as well as the concept) isn't settled. It's just a test so far.

**Reply by phill-85** [reply]
I've been thinking about it, and I agree with @jbtrystram

Though I'm for sticking with the PF guidelines, and following the React examples. If PF species that only certain children should be used then we should follow that. If there is requirements on the types of children then we shouldn't restrict that. I think this is especially important as the documentation of PF Yew pales is comparison to PF and PF React
